### ISSUE TITLE
Pre RFC / Feature Request: (De)structuring

### ISSUE THREAD

**Issue by NullVoxPopuli** [emberjs/rfcs#553]
As far as I know, there is no syntactic meaning in glimmer for `{ }` (single curlies).
So, without diving in at trying to implement this myself, it seems like we have room to implment something like the following:

```hbs
{{!-- Before --}}
<Button @variant='primary' as |yieldedHash|>
  <yieldedHash.Icon />
</Button>

{{!-- After --}}
<Button @variant='primary' as |{ Icon }|>
  <Icon />
</Button>
```
and
```hbs
{{!-- Before --}}
<Button @variant={{@variant}} @size={{@size}} @onClick={{this.onClick}}>
  {{@text}}
</Button>

{{!-- After -- I don't think this is the right way to do this, cause it's semantically weird --}}
<Button { @variant @size this.onClick }>
  {{@text}}
</Button>
```

Thoughts?

**Comment by lifeart** [comment]
```hbs
<Button @variant="primary" as | ( Icon Image ) ( Spinner ) |>
  <Icon />
  <Image />
  <Spinner />
</Button>
```

**Comment by NullVoxPopuli** [comment]
> `| ( Icon Image ) ( Spinner ) |`

I thought `()` already imply helper? 
hmm. idk. `as` is weird.

**Comment by lifeart** [comment]
or we can finally drop `{{`, `}}` and use `{`, `}`  :)

**Comment by NullVoxPopuli** [comment]
that would introduce ambiguity with what this question is about.
We want things _more_ clear, and easily differentiable, not less.

`{{}}` _always_ means mustache expression, which is (in strict):
 - value expression
   - which could be a helper invocation

to reduce `{{}}` to `{ }` means that we'd need to, at runtime, determine usage / intent. not ideal :(

**Comment by lifeart** [comment]
Svelte use mustache and  ‘{‘

**Comment by monovertex** [comment]
> ```
> {{!-- Before --}}
> <Button @variant={{@variant}} @size={{@size}} @onClick={{this.onClick}}>
>   {{@text}}
> </Button>
> 
> {{!-- After -- I don't think this is the right way to do this, cause it's semantically weird --}}
> <Button { @variant @size this.onClick }>
>   {{@text}}
> </Button>
> ```

We don't use angle bracket components in our project yet, so I'm not sure if this conflicts with anything, but what about dropping the curlies?

```handlebars
{{!-- After --}}
<Button @variant @size this.onClick>
  {{@text}}
</Button>
```

Basically the shorthand property value feature, just consider `<Button` and `>` as the opening and closing curlies.

**Comment by NullVoxPopuli** [comment]
> but what about dropping the curlies?

This already means positional params though.

**Comment by monovertex** [comment]
> This already means positional params though.

Right, my bad, forgot about positional params, we never use them for our components 😅

**Comment by ro0gr** [comment]
> This already means positional params though.

I don't think we have positional params for <AngleBrackets, do we?

**Comment by NullVoxPopuli** [comment]
> I don't think we have positional params for <AngleBrackets, do we?

Not with the Glimmer Component manager, but we can make component managers that support positional params. I've been wanting to dig in to this for control flow, for things like ember-animated, or other control-flow logic (maybe like with async-data loading at the component level for managing intermediate / placeholder state).

There is a write up on all that here: https://github.com/tildeio/ember-control-flow-component

**Comment by NullVoxPopuli** [comment]
yup, I was wrong about the positional thing in angle brackets:
![image](https://user-images.githubusercontent.com/199018/67572400-1d004e80-f704-11e9-8f61-3313c499f41c.png)

Thanks @rwjblue and @locks! <3

**Comment by locks** [comment]
> Svelte use mustache and ‘{‘

Svelte does not use mustache.

**Comment by lifeart** [comment]
@locks https://github.com/sveltejs/svelte/blob/master/src/compiler/parse/state/mustache.ts

**Comment by robclancy** [comment]
> > ```
> > {{!-- Before --}}
> > <Button @variant={{@variant}} @size={{@size}} @onClick={{this.onClick}}>
> >   {{@text}}
> > </Button>
> > 
> > {{!-- After -- I don't think this is the right way to do this, cause it's semantically weird --}}
> > <Button { @variant @size this.onClick }>
> >   {{@text}}
> > </Button>
> > ```
> 
> We don't use angle bracket components in our project yet, so I'm not sure if this conflicts with anything, but what about dropping the curlies?
> 
> ```
> {{!-- After --}}
> <Button @variant @size this.onClick>
>   {{@text}}
> </Button>
> ```
> 
> Basically the shorthand property value feature, just consider `<Button` and `>` as the opening and closing curlies.

This was my first thought when wanting to be able to remove the redundancy.

**Comment by wagenet** [comment]
Leaving this open for now since it's referenced in @NullVoxPopuli's meta issue.
